consensus/hotstuff/signature.go

```diff
@@ -8,37 +8,53 @@ import (
+	// TrustedAdd adds the signature share to the reconstructors internal
+	// state. Validity of signature is not checked. It is up to the
+	// implementation, whether it still adds a signature or not, when the
+	// minimal number of required sig shares has already been reached,
+	// because the reconstructed group signature is the same.
+	// Returns: true if and only if enough signature shares were collected
 	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
+
+	// HasSufficientShares returns true if and only if reconstructor
+	// has collected a sufficient number of signature shares.
 	HasSufficientShares() bool
-	// Reconstruct the beacon signature if there are enough shares. It errors if not enough shares were collected.
+
+	// Reconstruct reconstructs the group signature from the provided
+	// signature shares. Errors if the the number of shares is insufficient
+	// or some of the added signatures shares were invalid.
 	Reconstruct() (crypto.Signature, error)
 }
 
-// RandomBeaconAggregator aggregates the random beacon signatures
-type RandomBeaconAggregator interface {
-	// TrustedAdd adds an already verified signature.
-	// return (true, nil) means the signature has been added
-	// return (false, nil) means the signature is a duplication
-	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
-	// Aggregate assumes enough shares have been collected, it aggregates the signatures
-	// and return the aggregated signature.
-	// if called concurrently, only one thread will be running the aggregation.
-	Aggregate() ([]byte, error)
-}
+// SigType is the aggregable signature type.
+type SigType int
+
+// There are two signatures are aggregable, one is the normal staking signature,
+// the other is the threshold sig used as staking sigature.
+const (
+	SigTypeStaking SigType = iota
+	SigTypeRandomBeacon
+	SigTypeInvalid
+)
 
 // SignatureAggregator aggregates the verified signatures.
 // It can either be used to aggregate staking signatures or aggregate random beacon signatures,
 // but not a mix of staking signature and random beacon signature.
+// Implementation of SignatureAggregator must be concurrent safe
 type SignatureAggregator interface {
-	// TrustedAdd adds an already verified signature.
-	// return (true, nil) means the signature has been added
-	// return (false, nil) means the signature is a duplication
-	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
+	// TrustedAdd adds an already verified signature, and look up the weight for the given signer,
+	// and add it to the total weight, and returns the total weight that have been collected.
+	// return (true, 1000, nil) means the signature has been added, and 1000 weight has been collected in total.
+	// return (false, 1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
+	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (added bool, totalWeight uint64, exception error)
```

Comment:
shall we change `SignatureAggregator` to be the same as `RandomBeaconReconstructor` ? 
I mean we can also have a method for `HasSufficientWeight` and return true or false depending if we have collected enough weight

Reply:
I think we should just go with CombinedAggregator and track weight atomically for both things, it simplifies things a lot.

Reply:
The thing is when we receive a consensus vote, the signature could be either a staking sig or a random beacon sig. 

If it's a random beacon sig, we need to add the same random beacon sig to both random beacon aggregator and random beacon reconstructor atomically.

Meaning there should be only 2 cases: both added, none was added.
But not the following 2 cases:
- added only to random beacon aggregator, not in random beacon reconstructor
- added only to random beacon reconstructor, not in random beacon aggregator

I'm not sure how CombinedAggregator could ensure atomicity if the update operations happen in two separate modules: combined aggregator and random beacon constructor. Wouldn't we have to use a lock in VoteCollector anyway?

Reply:
I see this situation exactly as it is in [this](https://s3.us-west-2.amazonaws.com/secure.notion-static.com/759dfc32-a442-4610-b3fd-dda67ae14267/Concurrent_Vote_Processing_%286%29.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAT73L2G45O3KS52Y5%2F20210817%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20210817T142027Z&X-Amz-Expires=86400&X-Amz-Signature=10ab8406a1aa5938844ca76c1bc3763c54c79085aeead6bdd4198a57417b800a&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Concurrent_Vote_Processing_%286%29.png%22) diagram. 
1. We verify signature, we receive a signature type
2. Add it to combined aggregator
    2.1 lock self
    2.2. aggregate staking sig whenever it's a threshold or staking sig
    2.3 increase weight
    2.4 unlock
3. if it was a random beacon sig add it to reconstructor. 

This way we have two separate weights, one for aggregated signatures and one for random beacon shares.